File: apps/web/src/lib/captcha.ts

export async function verifyCaptcha({
  token,
  expectedAction,
  minScore = 0.5,
}: {
  token: string;
  expectedAction?: string;
  minScore?: number;
}): Promise<{ ok: boolean; data?: Record<string, unknown> }> {
  try {
    const secretKey = process.env.RECAPTCHA_SECRET_KEY;
    if (!secretKey) {
      return { ok: false, data: { error: "Server configuration error" } };
    }
    if (!token) return { ok: false, data: { error: "missing-token" } };

    const body = new URLSearchParams();
    body.set("secret", secretKey);
    body.set("response", token);

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 5000);

    let response: Response;
    try {
      response = await fetch(
        "https://www.google.com/recaptcha/api/siteverify",
        {
          method: "POST",
          headers: { "Content-Type": "application/x-www-form-urlencoded" },
          body: body.toString(),
          signal: controller.signal,
        },
      );
    } finally {
      clearTimeout(timeout);
    }

    const data = await response.json();

    const ok =
      data?.success === true &&
      (typeof data?.score !== "number" || data.score >= minScore) &&
      (!expectedAction || data.action === expectedAction);

    return { ok, data };
  } catch (error) {
    console.error("Error verifying captcha:", error);
    return { ok: false, data: { error: "exception" } };
  }
}

File: apps/api/src/captcha.rs
use serde::Deserialize;

use crate::Upstreams;

#[derive(Deserialize)]
struct SiteVerifyResponse {
    success: bool,
    score: Option<f64>,
    action: Option<String>,
    #[serde(default, rename = "error-codes")]
    error_codes: Vec<String>,
}

/// Server-side reCAPTCHA v3 verification.
///
/// Divergence from the old TypeScript implementation (which failed closed):
/// when `RECAPTCHA_SECRET_KEY` is unset the check passes with a warning, so
/// development and test environments work without Google. Production renders
/// the secret via Doppler; `main` logs loudly at startup when it is missing.
pub async fn verify(
    http: &reqwest::Client,
    upstreams: &Upstreams,
    token: &str,
    expected_action: &str,
    min_score: f64,
) -> bool {
    let Some(secret) = upstreams.recaptcha_secret.as_deref() else {
        tracing::warn!(
            action = expected_action,
            "RECAPTCHA_SECRET_KEY unset; skipping captcha verification"
        );
        return true;
    };
    if token.is_empty() {
        return false;
    }

    let url = format!("{}/recaptcha/api/siteverify", upstreams.recaptcha);
    let response = http
        .post(&url)
        .form(&[("secret", secret), ("response", token)])
        .timeout(std::time::Duration::from_secs(5))
        .send()
        .await;

    let body: SiteVerifyResponse = match response {
        Ok(res) => match res.json().await {
            Ok(body) => body,
            Err(err) => {
                tracing::error!(error = %err, "captcha response parse failed");
                return false;
            }
        },
        Err(err) => {
            tracing::error!(error = %err, "captcha verification request failed");
            return false;
        }
    };

    let score_passed = body.score.map(|s| s >= min_score).unwrap_or(true);
    let action_passed = body.action.as_deref() == Some(expected_action);
    let passed = body.success && score_passed && action_passed;
    if !passed {
        tracing::warn!(
            expected_action,
            actual_action = body.action.as_deref(),
            score = body.score,
            min_score,
            error_codes = ?body.error_codes,
            "captcha verification failed"
        );
    }

    passed
}
